### PR TITLE
Fix #5852: [java] UnnecessaryImport with multiple static imports from same class

### LINKED ISSUE #5852: [java] UnnecessaryImport with multiple static imports from same class
**Affects PMD Version:** 7.14.0

**Rule:** [UnnecessaryImport](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessaryimport)

**Description:**

**Code Sample demonstrating the issue:**

```
import static io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime;

[WARNING] PMD Failure: io.quarkus.runtime.configuration.ConfigRecorder:3 Rule:UnnecessaryImport Priority:4 Unused static import 'io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime'.

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.26.0:check (default-cli) on project quarkus-core: PMD 7.14.0 has found 1 violation. For more details see: /Users/vincent.potucek/IdeaProjects/quarkus/core/runtime/target/pmd.xml
```

```java
package io.quarkus.runtime.configuration;

import static io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime; // _not_ unnecessary!!
import static io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime.fail;
import static io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime.warn;

// ...
public class ConfigRecorder {
    public void handleConfigChange(Map<String, ConfigValue> buildTimeRuntimeValues) {
            // ...
            BuildTimeMismatchAtRuntime buildTimeMismatchAtRuntime = config
                    .getOptionalValue("quarkus.config.build-time-mismatch-at-runtime", BuildTimeMismatchAtRuntime.class)
                    .orElse(warn);
            if (fail.equals(buildTimeMismatchAtRuntime)) {
                throw new IllegalStateException(msg);
            } else if (warn.equals(buildTimeMismatchAtRuntime)) {
                log.warn(msg);
            }
    }
}
```

<details>
  <summary>Full source</summary>

```java
package io.quarkus.runtime.configuration;

import static io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime;
import static io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime.fail;
import static io.quarkus.runtime.ConfigConfig.BuildTimeMismatchAtRuntime.warn;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

import org.eclipse.microprofile.config.ConfigProvider;
import org.eclipse.microprofile.config.spi.ConfigSource;
import org.jboss.logging.Logger;

import io.quarkus.runtime.ShutdownContext;
import io.quarkus.runtime.annotations.Recorder;
import io.smallrye.config.ConfigValue;
import io.smallrye.config.SmallRyeConfig;

@Recorder
public class ConfigRecorder {
    private static final Logger log = Logger.getLogger(ConfigRecorder.class);

    public void handleConfigChange(Map<String, ConfigValue> buildTimeRuntimeValues) {
        SmallRyeConfig config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
        // Disable the BuildTime RunTime Fixed (has the highest ordinal), because a lookup will get the expected value,
        // and we have no idea if the user tried to override it in another source.
        Optional<ConfigSource> builtTimeRunTimeFixedConfigSource = config.getConfigSource("BuildTime RunTime Fixed");
        if (builtTimeRunTimeFixedConfigSource.isPresent()) {
            ConfigSource configSource = builtTimeRunTimeFixedConfigSource.get();
            if (configSource instanceof DisableableConfigSource) {
                ((DisableableConfigSource) configSource).disable();
            }
        }

        List<String> mismatches = new ArrayList<>();
        for (Map.Entry<String, ConfigValue> entry : buildTimeRuntimeValues.entrySet()) {
            ConfigValue currentValue = config.getConfigValue(entry.getKey());
            // Check for changes. Also, we only have a change if the source ordinal is higher
            // The config value can be null (for ex. if the property uses environment variables not available at build time)
            if (currentValue.getValue() != null && !Objects.equals(entry.getValue().getValue(), currentValue.getValue())
                    && entry.getValue().getSourceOrdinal() < currentValue.getSourceOrdinal()) {
                mismatches.add(
                        " - " + entry.getKey() + " is set to '" + currentValue.getValue()
                                + "' but it is build time fixed to '"
                                + entry.getValue().getValue() + "'. Did you change the property " + entry.getKey()
                                + " after building the application?");
            }
        }

        // Enable the BuildTime RunTime Fixed. It should be fine doing these operations, because this is on startup
        if (builtTimeRunTimeFixedConfigSource.isPresent()) {
            ConfigSource configSource = builtTimeRunTimeFixedConfigSource.get();
            if (configSource instanceof DisableableConfigSource) {
                ((DisableableConfigSource) configSource).enable();
            }
        }

        if (!mismatches.isEmpty()) {
            String msg = "Build time property cannot be changed at runtime:\n" + String.join("\n", mismatches);
            // TODO - This should use ConfigConfig, but for some reason, the test fails sometimes with mapping not found when looking ConfigConfig
            BuildTimeMismatchAtRuntime buildTimeMismatchAtRuntime = config
                    .getOptionalValue("quarkus.config.build-time-mismatch-at-runtime", BuildTimeMismatchAtRuntime.class)
                    .orElse(warn);
            if (fail.equals(buildTimeMismatchAtRuntime)) {
                throw new IllegalStateException(msg);
            } else if (warn.equals(buildTimeMismatchAtRuntime)) {
                log.warn(msg);
            }
        }
    }

    public void handleNativeProfileChange(List<String> buildProfiles) {
        SmallRyeConfig config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
        List<String> runtimeProfiles = config.getProfiles();

        if (buildProfiles.size() != runtimeProfiles.size()) {
            log.warn(
                    "The profile '" + buildProfiles + "' used to build the native image is different from the runtime profile '"
                            + runtimeProfiles + "'. This may lead to unexpected results.");
            return;
        }

        for (int i = 0; i < buildProfiles.size(); i++) {
            String buildProfile = buildProfiles.get(i);
            String runtimeProfile = runtimeProfiles.get(i);

            if (!buildProfile.equals(runtimeProfile)) {
                log.warn("The profile '" + buildProfile
                        + "' used to build the native image is different from the runtime profile '" + runtimeProfile
                        + "'. This may lead to unexpected results.");
            }
        }
    }

    public void unknownConfigFiles() throws Exception {
        ConfigDiagnostic.unknownConfigFiles(ConfigDiagnostic.configFilesFromLocations());
    }

    public void releaseConfig(ShutdownContext shutdownContext) {
        // This is mostly useful to handle restarts in Dev/Test mode.
        // While this may seem to duplicate code in IsolatedDevModeMain,
        // it actually does not because it operates on a different instance
        // of QuarkusConfigFactory from a different classloader.

        if (shutdownContext == null) {
            throw new RuntimeException(
                    "Internal errror: shutdownContext is null. This probably happened because Quarkus failed to start properly in an earlier step, or because tests were run on a Quarkus instance that had already been shut down.");
        }
        shutdownContext.addLastShutdownTask(QuarkusConfigFactory::releaseTCCLConfig);
    }
}
```

</details>

- https://github.com/quarkusio/quarkus/pull/48414/files

**Expected outcome:**

PMD reports a violation at line 3, but that's wrong. That's a false positive.

**Running PMD through:** *[Maven]*

### PR DESCRIPTION
**Pull request by Pankraz76**
## Related issues

- Fix #5852

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by Pankraz76** [comment]
added test, could not fix impl. myself. Feel free to take over @oowekyala.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 1]
```diff
@@ -1474,6 +1474,126 @@ public class Foo {
      */
     public void test() {}
 
+}
+        ]]></code>
+    </test-code>
+    <test-code>
+        <description>#5852: [java] UnnecessaryImport with multiple static imports from same class</description>
```
Comment (Pankraz76) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml]: issue: 

image

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryImportRule.java [hunk 2]
```diff
@@ -368,10 +368,7 @@ private <T> void checkScopeChain(boolean recursive,
 
                 if (scopeIter.getScopeTag() == ScopeInfo.SINGLE_IMPORT) {
 
-                    allSingleNameImports.removeIf(
-                        it -> (it.isStatic() || !onlyStatic)
-                            && symbol.getSimpleName().equals(it.node.getImportedSimpleName())
```
Comment (Pankraz76) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryImportRule.java:373]: this one is untested and fixes the issue, but imposed another regarding the same concern.

**Review by Pankraz76 — COMMENTED** [review]
issue ready to investigate by someone in touch with the impl.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 3]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+                        } else if (warn.equals(buildTimeMismatchAtRuntime)) {
+                            log.warn(msg);
+                        }
+                    }
+                }
+            }
+        ]]></code>
+    </test-code>
+<!--    <test-code>-->
+<!--        <description>#5852: [java] UnnecessaryImport with multiple static imports from same class - true positive</description>-->
```
Comment (Pankraz76) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1514]: uncomment to see new issue.

**Comment by Pankraz76** [comment]
Thanks @iddeepak for the contribution on 
- https://github.com/google/google-java-format/pull/1266

You pushed the code so far, the same success here would fix it all, thus the PMD appreciate community your efforts.

The tests are working well, debug is easy.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 4]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+        <description>Static import of nested class</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import static java.util.Map.Entry;
+public class Foo {
+    Entry<String, String> entry;
+}
+        ]]></code>
+    </test-code>
+
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1562]: Are these test cases here below really part of your PR? Or are they duplicated now? I remember seeing this on another fix (#5818) - ok, not exactly.
I'm just wondering, what you did here. It looks like they have nothing to do with the issue being fixed (#5852), but add additional test cases, that were not covered yet?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 5]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+    <!-- Test cases for imports used in module-info.java -->
+<!--    <test-code>-->
+<!--        <description>Import used in module-info requires</description>-->
+<!--        <expected-problems>0</expected-problems>-->
+<!--        <code><![CDATA[-->
+<!--import some.module.ModuleName;-->
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1712]: This doesn't sound correct, you cannot import modules that way. What you can do is, to import on demand all packages that a module exports via `import module ...` (https://openjdk.org/jeps/476). But you wouldn't use that in a module-info.java file...

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml (hunk 4 above)
Comment (Pankraz76) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1562]: > , that were not covered yet?

yes i consulted some guy to check for missing test cases. That the result.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 6]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+<!--module my.module {-->
+<!--    requires ModuleName;-->
+<!--}-->
+<!--        ]]></code>-->
+<!--    </test-code>-->
```
Comment (Pankraz76) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1717]:
```suggestion
```
imho dont invest on edge case.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 7]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
      */
     public void test() {}
 
+}
+        ]]></code>
+    </test-code>
+    <test-code>
+        <description>#5852: [java] UnnecessaryImport with multiple static imports from same class - false negative</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1481]: I think, we should add another test case. Your test case tests the behavior of PMD, if it is not used correctly - it is missing quarkus on the auxclass path - which means, we don't resolve the types and e.g. can't distinguish whether `ConfigConfig` is a package name or a class (I assume, it is class because of naming conventions).

That's why you need to really reproduce the case correctly:
* Create a class `ConfigConfig` in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport`
* This class should have the nested static class `Foo` which itself has two public static fields `fail` and `warn`.
* Maybe you can simplify the code a bit to get rid of unnecessary code lines which are irrelevant for this bug (e.g. no logger is required).
* In the end, you code sample should compile - this one doesn't...

Without a proper test, I'm not convinced, that the fix fixes the problem.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml (hunk 3 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1514]: ok, then it seems, the fix is not finished yet?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml (hunk 4 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1562]: ok, thanks for adding additional test cases.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 8]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in annotations -->
+    <test-code>
+        <description>Import used in annotation value</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.MyEnum;
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1579]: This test case should be improved: Please create the related classes/types in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport` so that we actually use type resolution correctly.

We probably also need to import `MyAnnotation`.

Alternatively, you might find in the standard java api an enum and an anntoation, you can reuse in the test, so that you don't need to create your own types.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 9]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+@MyAnnotation(MyEnum.VALUE)
+public class Foo {}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Import used in array annotation value</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.MyEnum;
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1589]: This test case should be improved: Please create the related classes/types in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport` so that we actually use type resolution correctly.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 10]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+
+    <!-- Test cases for imports used in lambda parameters -->
+    <test-code>
+        <description>Import used in lambda parameter type</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.io.File;
+public class Foo {
+    void test() {
+        list.stream().map((File f) -> f.getName());
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1603]: this won't compile: `list` is not known here, so PMD won't know, what the method `stream()` will return etc...

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 11]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+
+    <!-- Test cases for imports used in method references -->
+    <test-code>
+        <description>Import used in method reference</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.io.File;
+public class Foo {
+    void test() {
+        list.stream().map(File::getName);
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1617]: this won't compile: `list` is not known here, so PMD won't know, what the method `stream()` will return etc...

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 12]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in switch expressions -->
+    <test-code>
+        <description>Import used in switch expression</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.MyEnum;
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1628]: This test case should be improved: Please create the related classes/types in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport` so that we actually use type resolution correctly.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 13]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+public class Foo {
+    String test(MyEnum e) {
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1630]: I'm not sure, if this test case tests what you intended: The type `MyEnum` is used here as the method parameter, so the import is for this line required.
The cases in the switch expression are unrelated and could be removed.

Maybe you want to test the reverse, e.g. the `import static some.pkg.MyEnum.VALUE1` is unnecessary?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 14]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+public record Person(String name, LocalDate birthDate) {}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in sealed classes -->
+    <test-code>
+        <description>Import used in sealed class hierarchy</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.Shape;
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1655]: This test case should be improved: Please create the related classes/types in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport` so that we actually use type resolution correctly.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 15]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+        <code><![CDATA[
+import java.time.LocalDate;
+public class Foo {
+    String test(String name) {
+        return """
+            Hello %s!
+            Today is %s
+            """.formatted(name, LocalDate.now());
+    }
 }
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1689]: Mabye this test case could be simplified? Text block is a string and text block is irrelevant here.

You can achieve the same effect with that:

```java
import java.time.LocalDate;
public class Foo {
  {
     LocalDate.now();
  }
}
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml (hunk 5 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1712]: You indeed can have some imports in a module-info file, see https://docs.oracle.com/javase/specs/jls/se21/html/jls-7.html#jls-7.3

If the module consumes a service, you reference a type. And you can then import this type, so that you don't need to use the fully qualified name and can use the simple name instead, e.g.

```java
import net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport.ServiceInterface;
import net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport.ServiceImpl;

module my.module {
    uses ServiceInterface;
    provides ServiceInterface with ServiceImpl;
}
```

see also https://docs.oracle.com/javase/specs/jls/se21/html/jls-7.html#jls-7.7

but you can't shorten a module name - module names always need to be fully qualified.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 16]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+<!--}-->
+<!--        ]]></code>-->
+<!--    </test-code>-->
+
+    <!-- Test cases for imports used in Java 14+ record patterns -->
+    <test-code>
+        <description>Import used in record pattern</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.Point;
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1724]: This test case should be improved: Please create the related classes/types in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport` so that we actually use type resolution correctly.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 17]
```diff
@@ -1474,8 +1474,281 @@ public class Foo {
...
+}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in Java 22+ unnamed patterns/variables -->
+    <test-code>
+        <description>Import used with unnamed pattern variable</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.Point;
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:1740]: This test case should be improved: Please create the related classes/types in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport` so that we actually use type resolution correctly.

**Comment by Pankraz76** [comment]
closing to stop starting start finishing.
